proposed/2021/MachineReadableOutput.md

```diff
@@ -0,0 +1,797 @@
+
+   Transitive Package                                                                   Resolved
+   > Microsoft.CSharp                                                                   4.0.1
+   > Microsoft.NETCore.Platforms                                                        1.1.0
+   > Microsoft.NETCore.Targets                                                          1.1.0
+...
+
+Project 'MyProjectB' has the following package references
+   [netcoreapp3.1]:
+   Top-level Package      Requested             Resolved
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
+   > Text2Xml.Lib         1.1.2                 1.1.2
+
+   Transitive Package                                                                   Resolved
+   > Microsoft.CSharp                                                                   4.0.1
+   > Microsoft.NETCore.Platforms                                                        1.1.0
+   > Microsoft.NETCore.Targets                                                          1.1.0
+   > Microsoft.Win32.Primitives                                                         4.3.0
+...
+
+   [net5.0]:
+   Top-level Package      Requested             Resolved
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
+   > Text2Xml.Lib         1.1.2                 1.1.2
+
+   Transitive Package                                                                   Resolved
+   > Microsoft.CSharp                                                                   4.0.1
+   > Microsoft.NETCore.Platforms                                                        1.1.0
+   > Microsoft.NETCore.Targets                                                          1.1.0
+...
+
+```
+
+### dotnet list package --include-transitive --json --flatten
+
+```json
+{
+  "version": 1,
+  "topLevelPackages": [
+       {
+            "id": "Microsoft.Extensions.Primitives",
+            "requestedVersion": "[1.0.0, 5.0.0]",
+            "resolvedVersion": "1.0.0"
+        },
+        {
+            "id": "NuGet.Commands",
+            "requestedVersion": "4.8.0-preview3.5278",
+            "resolvedVersion": "4.8.0-preview3.5278"
+        },
+        {
+            "id": "Text2Xml.Lib",
+            "requestedVersion": "[1.1.2, 2.0.0)",
+            "resolvedVersion": "1.1.2"
+        }
+    ],
+    "transitivePackages": [
+        {
+            "id": "Microsoft.CSharp",
+            "resolvedVersion": "4.0.1"
+        },
+        {
+            "id": "Microsoft.NETCore.Platforms",
+            "resolvedVersion": "1.1.0"
+        },
+        {
+            "id": "Microsoft.NETCore.Targets",
+            "resolvedVersion": "1.1.0"
+        }
+...
+    ]
+}
+```
+
+### Compatibility
+
+ We start with `version 1`, as long as we don't remove or rename then it'll be backward compatible. In case [we change version](https://stackoverflow.com/a/13945074) just add new properties, keep old ones even it's not used.
```

Comment:
> We start with `version 1`, as long as we don't remove or rename then it'll be backward compatible.

In that case, should we provide a `--json-version 1` or `--output-version 1` option, so if we create a version 2 output, then customers who are writing scripts to parse the json from version 1 don't suddenly start breaking?  Remember that customers using cloud hosted CI agents may not choose when the version of the dotnet sdk is upgraded. Ok, there is the `dotnet-install.[ps1|sh]` script, allowing customers to install specific versions they want, and Azure DevOps does have a `UseDotNet` task to select a specific version. Is our guidance that customers who want deterministic output, that they must also control which specific version of the dotnet sdk is being used, or should we give customers the ability to control which version NuGet outputs?

> keep old ones even it's not used

That doesn't feel like a very honest definition of backwards compatibility. If customers write scripts to parse the json and use the property, then a new version of the dotnet cli that starts output empty values is going to break the customer script's business logic. Sure, from a schema point of view it's still there, but from a pragmatic point of view it's breaking.

Reply:
Versioning the output seems odd to me, `vswhere` and `npm ls` don't have this.  Do we need to?  Let's just never break the schema...

Reply:
In which case doing good design up-front is more important, and we should leave the design spec up for a longer period of time and encourage customers to provide feedback, not treat this is a "quick" customer sprint issue where we rush the design and implementation.

There are trade-offs either way, we have to choose which one is more important to us.

Reply:
especially when half the team is on vacation at the moment

Reply:
> In which case doing good design up-front is more important, and we should leave the design spec up for a longer period of time and encourage customers to provide feedback, not treat this is a "quick" customer sprint issue where we rush the design and implementation.
> 
> There are trade-offs either way, we have to choose which one is more important to us.

I'm favor of just having `version`. But not sure about `--output-version 1` part, as long as we're additive we're good. Of course, I welcome more feedbacks. Implementing this is not really hard, but I feel we should focus on MVP.

Reply:
Another feedback related to this. https://github.com/NuGet/Home/pull/11446#discussion_r766776940

Reply:
> as long as we're additive we're good

In which case version is completely unnecessary.  But if we're saying "we're not 100% sure things won't change in 10 years from now, so let's add a version property to the output just to be safe", then by using the same logic we should allow the customer to choose what version of the output they want, assuming the version of the tool supports the output version requested.

> I feel we should focus on MVP.

Consider the customer impact. We don't add a `--output-version` now because we think that this is what agile software development and MVP's mean 🙄 . Customers write scripts to use this in their CI.  In a year or longer we decide we want to make a breaking change, but also implement `--output-version` so customers who have existing scripts aren't forced to change yet. However, since all these customers wrote scripts before `--output-version` was implemented, none of their scripts use it.  And since the vast majority of people parse json in a non-strict mode, when the default output changes, their scripts are not going to fail, notifying them that something change changed. No, their scripts will fail to detect that the projects are using packages with known vulnerabilities.

Not doing things in order to start with an MVP and add features later works only when it's something that won't cause problems when added later, unless it's something that's intentionally a breaking change. But the entire point of the `version` property in the json and/or an `--output-version` argument is to provide a safe way to make breaking changes.

Reply:
During my discussion with Jeff I noticed 1 point: Versioning’s mostly suitable for file format which read/write by app, but for files which can be edited by human manually having explicit versioning doesn’t make much sense. So, I'm going to include versioning here. By default, we output current/latest schema version (which is 1 for us) output, same as most other application does. But `--json-version 1` support is up to debate if we really come up with `version 2` schema in the future, by then we can make much educated decision.

Reply:
Please consider doing a cross-comparison of other CLI experiences and whether they include schema versions in machine readable output.

Reply:
Unfortunately, there are not much other .net CLI experience to compare. Only 1 which output is `dotnet counter` but output is very simply <key value pair>. By the way we already have explicit versioning in `.nupkg.metadata` (version2), `packages.lock.jsons` file (version 1), I don't see any reason we do something different in this case

Reply:
I meant in other ecosystems that list out dependency graphs. There have been some hinted already such as `npm ls, `go list`, `cargo tree`, and more.

Reply:
> I meant in other ecosystems that list out dependency graphs. There have been some hinted already such as `npm ls`, `go list`, `cargo tree`, and more.

I already added my observation of `npm ls json` into prior art section, it's quite simpler than ours and doesn't have versioning.  
For below tools I never used before I'm just sharing what I have found from search engine.
[go list](https://dave.cheney.net/2014/09/14/go-list-your-swiss-army-knife) looks like `npm ls --parseable` output than json output.
[cargo tree](https://lib.rs/crates/cargo-tree) is more of graphic tool than readable output.

Reply:
In addition to our `.nupkg.metadata`, `packages.lock.json`, we have versioning [dependencyGraph ](https://gist.github.com/erdembayar/351e247185519d169c6afc4077007274) file too.

Reply:
After team discussion we include `versioning` in schema and output versioning like below.

```
dotnet list package --format json --output-version 1
dotnet list package --format yaml --output-version 1
```